Add a .NET Core test project for FakeItEasy.Specs

thomaslevesque opened the pull request:
Connects to #531 

Well, it wasn't as painful as I feared it would be :wink:

thomaslevesque commented on FakeItEasy.sln:
```diff
@@ -10,13 +10,13 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 		FakeItEasy.snk = FakeItEasy.snk
 		Gemfile = Gemfile
 		Gemfile.lock = Gemfile.lock
-		cacert.pem = cacert.pem
```
I didn't intend to reorder those lines, but I guess it can't hurt, and there's no point if fighting VS over this.

thomaslevesque commented on tests/FakeItEasy.Specs.netstd/ICloneable.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace FakeItEasy.Specs
+{
+    public interface ICloneable
```
This type doesn't exist in .NET Standard. The `Clone` method isn't actually necessary, but I included it for consistency with the BCL type.

blairconrad replied:
An alternative would be to just use a different interface in `CreationOptionsSpecs.cs` and `FakeOptionsBuilderSpecs.cs`, but this is good.
We could always change interfaces later, if we're so inclined.

thomaslevesque commented on tests/FakeItEasy.Specs.netstd/project.json:
```diff
@@ -0,0 +1,69 @@
+        "System.Threading.Tasks.Parallel": "4.0.1",
+        "Microsoft.CSharp": "4.0.1",
+        "dotnet-test-xunit": "2.2.0-preview2-build1029",
+        "Xbehave.Core": "2.2.0-beta0001-build680"
```
Some of these dependencies are probably not necessary; is there an easy way to detect unused dependencies?

blairconrad replied:
I'm gonna run off and try this on the integration tests now!

thomaslevesque commented on tests/FakeItEasy.Specs/CallMatchingSpecs.cs:
```diff
@@ -227,10 +227,10 @@ public interface IHaveAnOutParameter
 
   Assertion failed for the following call:
-    System.Collections.Generic.IDictionary`2[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].TryGetValue(""any key"", <out parameter>)
+    System.Collections.Generic.IDictionary`2[[System.String, *, Version=4.0.0.0, Culture=neutral, PublicKeyToken=*],[System.String, *, Version=4.0.0.0, Culture=neutral, PublicKeyToken=*]].TryGetValue(""any key"", <out parameter>)
```
.NET Core has System.Private.CorLib instead of mscorlib, and a different PK token

thomaslevesque commented on tests/FakeItEasy.Specs/FakingInternalsSpecs.cs:
```diff
@@ -22,11 +22,9 @@ public static class FakingInternalsSpecs
-                    });
+                .x(() => exception.Message.Should()
+                    .Contain("Make it public, or internal and mark your assembly with")
+                    .And.Match("*[assembly: InternalsVisibleTo(\"DynamicProxyGenAssembly2*\")]*"));
```
The .NET Core version of Castle.Core uses the fully qualified assembly name

thomaslevesque commented on tests/FakeItEasy.Specs.netstd/project.json:
```diff
@@ -0,0 +1,69 @@
+      ],
+      "exclude": [
+        "../FakeItEasy.Specs/SelfInitializedFakesSpecs.cs",
+        "../FakeItEasy.Specs/DisposableSpecs.cs"
```
Oh, I forgot to mention this. These specs rely on `AppDomain.UnhandledException`, which isn't available in .NET Standard. There's no replacement yet, but it should be added in .NET Core 1.2.

blairconrad commented on tests/FakeItEasy.Specs.netstd/project.json:
```diff
@@ -0,0 +1,69 @@
+        "../FakeItEasy.Specs/DisposableSpecs.cs"
+      ]
+    },
+    "define": [ "NETCORE" ]
```
I just tried without this line, and the project builds and tests pass.

blairconrad commented:
@thomaslevesque, looks very good. Thanks.

My only substantive comments are around the redundancies in the project.json (some of which you pointed out), which you inherited from my integration tests efforts.
Oh, and the cacert.pem reordering. If you'd rather take care of that work here, #929 can be closed.

thomaslevesque commented:
I rebased, fixed the project.json to be similar to the one from FakeItEasy.IntegrationTests.netstd, and removed the `ICloneable` interface. Let me know when I can squash the fixup commit.

blairconrad commented:
Thanks, @thomaslevesque. I think it looks great!

adamralph commented:
Thanks @thomaslevesque, this looks ready to squash.

thomaslevesque commented:
squashed

blairconrad commented:
Thanks, @thomaslevesque! It's a brave new world.
